### DISCUSSION TITLE
Session.execute doesn't always autoflush?

### DISCUSSION THREAD

**Discussion by dsedivec** [sqlalchemy/sqlalchemy#9776] [Usage Questions] [answered]
Greetings!  The [SQLAlchemy docs on autoflush][1] say:

> Autoflush is defined as a configurable, automatic flush call which occurs at the beginning of methods including:
>
> * Session.execute() and other SQL-executing methods

[1]: https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-flushing

However, this confused me when the use of `execute` with a literal SQL query doesn't seem to autoflush.  Here's a tiny example:

```python
import sqlalchemy as sa
from sqlalchemy import orm


class MyBase(orm.DeclarativeBase):
    pass


class User(MyBase):
    __tablename__ = "user"

    name: orm.Mapped[str] = orm.mapped_column(sa.String(50), primary_key=True)


engine = sa.create_engine("sqlite://")
MyBase.metadata.create_all(engine)

with orm.Session(engine) as sess:
    sess.add(User(name="alice"))

    # Here's the Session.execute with a SQL string:
    before_raw_results = sess.execute(sa.text("select * from user")).fetchall()
    print(
        "Table is unexpectedly empty, execute didn't autoflush:",
        repr(before_raw_results),
    )

    # What I'll call "ORM uses of execute" do autoflush:
    orm_results = sess.execute(sa.select(User)).fetchall()
    assert len(orm_results) == 1

    # Just confirming that autoflush happened, direct
    # SQL query sees the row in the DB now:
    after_raw_results = sess.execute(sa.text("select * from user")).fetchall()
    print("Autoflush happened, table now has rows:", repr(after_raw_results))
```

When you run this, you get (tested under Python 3.11 on macOS):

```
Table is unexpectedly empty, execute didn't autoflush: []
Autoflush happened, table now has rows: [('alice',)]
```

I bet what's happening here is that the docs meant something like "*normal ORM uses* of `Session.execute` will autoflush", rather than my initial reading of "*literally all uses* of `Session.execute` will autoflush"?  If that's right, would you like me to submit a docs PR?

Alternatively, maybe someone can point out where my error is. :)

Thank you, tireless SQLAlchemy maintainer(s)!

**Comment by zzzeek** [comment, marked as answer]
> Greetings! The [SQLAlchemy docs on autoflush](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-flushing) say:
> 
> > Autoflush is defined as a configurable, automatic flush call which occurs at the beginning of methods including:
> > 
> > * Session.execute() and other SQL-executing methods
> 
> However, this confused me when the use of `execute` with a literal SQL query doesn't seem to autoflush. Here's a tiny example:

the doc is wrong.   only execute() calls against constructs that refer to ORM entities, so literal text is not included.  Will update the doc now.

> I bet what's happening here is that the docs meant something like "_normal ORM uses_ of `Session.execute` will autoflush", rather than my initial reading of "_literally all uses_ of `Session.execute` will autoflush"? If that's right, would you like me to submit a docs PR?

That's the general idea, but this is specific enough that I'll just update it, thanks
